### PR TITLE
Add display of internal circuits for ControlFlowOps to text drawer

### PR DESCRIPTION
**Pull request by enavarro51**
### Summary

Displays the circuits inside ControlFlowOps instead of just displaying a box.

### Details and comments

This PR adds displays of internal circuits for ControlFlowOps to the circuit text drawer. This is similar to the capability in the mpl drawer in #10207. The main differences are,
- In order to not recreate the wheel, the basic structure of the text drawer was maintained. Doing this meant it would have been difficult to draw complete boxes around the circuits and have them wrap onto multiple lines in the way they were done in the mpl drawer. Instead the flow circuits are bracketed at the beginning and end. See examples below.
- When nesting control flow ops inside others, the mpl drawer used colors to indicate where the nesting was taking place. For the text drawer, a nesting number was used. So `If-1`, `End-1` would appear inside `If-0`, `End-0`.

Otherwise all the previous functionality of the text drawer is intact. This PR incorporates the changes to `utils.py` that were made in #10207.

- [x] Add tests for text drawer
- [x] Release note

```
qr = QuantumRegister(5, "q")
cr = ClassicalRegister(3, "cr")
qc = QuantumCircuit(qr, cr)

qc.h(0)
qc.h(1)

with qc.if_test((cr, 1)):
    qc.y(1)
    qc.x(2)
    qc.z(3)

      ┌───┐                       
 q_0: ┤ H ├───────────────────────
      ├───┤┌─────  ┌───┐  ──────┐ 
 q_1: ┤ H ├┤       ┤ Y ├        ├─
      └───┘│       ├───┤        │ 
 q_2: ─────┤ If-0  ┤ X ├  End-0 ├─
           │       ├───┤        │ 
 q_3: ─────┤       ┤ Z ├        ├─
           └──╥──  └───┘  ──────┘ 
 q_4: ────────╫───────────────────
              ║                   
cr_0: ════════■═══════════════════
              ║                   
cr_1: ════════o═══════════════════
              ║                   
cr_2: ════════o═══════════════════
             0x1                  
```
```
qr = QuantumRegister(4, "q")
cr = ClassicalRegister(3, "cr")
qc = QuantumCircuit(qr, cr)

a = Parameter("a")
qc.h(0)
qc.measure(0, 2)
with qc.for_loop((2, 4, 8, 16), loop_parameter=a):
    qc.h(0)
    qc.cx(0, 1)
    qc.rx(pi / a, 1)
    qc.measure(0, 0)
    with qc.if_test((cr[2], 1)):
        qc.z(0)
qc.draw("text")

      ┌───┐┌─┐┌────────────────────  ┌───┐                ┌─┐┌────── ┌───┐»
 q_0: ┤ H ├┤M├┤                      ┤ H ├──■─────────────┤M├┤ If-1  ┤ Z ├»
      └───┘└╥┘│ For-0 (2, 4, 8, 16)  └───┘┌─┴─┐┌─────────┐└╥┘└──╥─── └───┘»
 q_1: ──────╫─┤                      ─────┤ X ├┤ Rx(π/a) ├─╫────╫─────────»
            ║ └────────────────────       └───┘└─────────┘ ║    ║         »
 q_2: ──────╫──────────────────────────────────────────────╫────╫─────────»
            ║                                              ║    ║         »
 q_3: ──────╫──────────────────────────────────────────────╫────╫─────────»
            ║                                              ║    ║         »
cr_0: ══════╬══════════════════════════════════════════════╩════╬═════════»
            ║                                                   ║         »
cr_1: ══════╬═══════════════════════════════════════════════════╬═════════»
            ║                                                   ║         »
cr_2: ══════╩═══════════════════════════════════════════════════■═════════»
                                                                          »
«       ───────┐   ──────┐ 
« q_0:   End-1 ├─        ├─
«       ───────┘   End-0 │ 
« q_1: ──────────        ├─
«                  ──────┘ 
« q_2: ────────────────────
«                          
« q_3: ────────────────────
«                          
«cr_0: ════════════════════
«                          
«cr_1: ════════════════════
«                          
«cr_2: ════════════════════
«                          
```

### CODE REVIEW

**Comment by enavarro51** [comment]
@jakelishman I updated the text drawer to deal with getting the circuits from `node.op.blocks` and proper usage of the `inner_wire_map` so it works as the mpl drawer does. https://github.com/Qiskit/qiskit-terra/pull/10414/commits/cc8b7b79ada0eb068ae07f028b479a8e69aecad8

**Comment by enavarro51** [comment]
Update: The #10601 issue for the mpl drawer has now been fixed for the text drawer as well. An additional test was added.

**Comment by enavarro51** [comment]
@jakelishman This latest update incorporates the changes to the mpl drawer in #10602. This included expanding the use of the wire_map and refactoring the conditional section.

File: qiskit/visualization/circuit/text.py [hunk 1]
```diff
@@ -37,9 +36,15 @@
...
     get_condition_label_val,
+    _get_layered_instructions,
 )
 from ..exceptions import VisualizationError
 
+# Indicators for left, middle, and right of control flow gates
+CF_LEFT = 0
+CF_MID = 1
+CF_RIGHT = 2
+
```
Comment (jakelishman) [inline qiskit/visualization/circuit/text.py:47]: Python's got a standard-library type `enum.Enum` for things like this in general, but it's probably not worth bothering changing here.

File: qiskit/visualization/circuit/text.py [hunk 2]
```diff
@@ -317,6 +322,103 @@ def __init__(self, label, input_length, bot_connect=None, wire_label="", conditi
             self.top_connect = label
 
 
+class FlowOnQuWire(DrawElement):
+    """Draws a box for a ControlFlowOp using a single qubit."""
+
+    def __init__(self, label="", top_connect="─", conditional=False, section=CF_LEFT):
```
Comment (jakelishman) [inline qiskit/visualization/circuit/text.py]: Minor style: here and in the other new classes, it feels a bit weird to have `section` be optional; I don't think there's any default value that makes sense and the outer context should always known whether it's left/mid/right.

File: qiskit/visualization/circuit/text.py [hunk 3]
```diff
@@ -317,6 +322,103 @@ def __init__(self, label, input_length, bot_connect=None, wire_label="", conditi
...
+        super().__init__(label)
+        if section == CF_RIGHT:
+            self.top_format = " ─%s─┐"
+            self.mid_format = "  %s ├"
+            self.bot_format = " ─%s─┘"
+        else:
+            self.top_format = "┌─%s─ "
+            self.mid_format = "┤ %s  "
+            self.bot_format = "└─%s─ "
```
Comment (jakelishman) [inline qiskit/visualization/circuit/text.py:337]: Minor: how do you think it'd look if we had "mid" boxes display neither left nor right edge, so they appear more like a full continuation?  At the moment it looks a shade unbalanced to me that we get more opening "brackets" than closing ones for if/else and switch constructs.

File: qiskit/visualization/circuit/text.py [hunk 4]
```diff
@@ -673,16 +782,28 @@ def single_string(self):
         Returns:
             str: The lines joined by a newline (``\\n``)
         """
+        # Because jupyter calls both __repr__ and __repr_html__, this prevents the code
+        # from running twice.
+        if self._single_string:
+            return self._single_string
+        else:
+            # In case there is an exception, this prevents it from printing twice.
+            self._single_string = " "
```
Comment (jakelishman) [inline qiskit/visualization/circuit/text.py]: Tbh I'd probably skip this; these exceptions aren't actionable by a user, and an exception here would always indicate either an invalid circuit or a problem in our code, and we don't need to be too worried about optimising the return speed in those cases.  I'm worried that setting to the empty string implies there's some meaning to it.

File: test/python/visualization/test_circuit_text_drawer.py [hunk 5]
```diff
@@ -319,21 +320,30 @@ def test_wire_order(self):
...
+                "               ║  ",
+                " c_1: 0 ═══════■══",
+                "               ║  ",
+                " c_0: 0 ═══════o══",
+                "               ║  ",
+                " c_3: 0 ═══════■══",
+                "              0xa ",
             ]
         )
+
```
Comment (jakelishman) [inline test/python/visualization/test_circuit_text_drawer.py:346]: This test of `wire_order` looks unrelated; how come it needs changing?

**Review by jakelishman — COMMENTED** [review]
I'm super super excited to have this in, thanks so much for working on it!

We spoke about this offline, but for posterity for others: I noticed this edge case produces unexpected output for me - an empty control-flow op doesn't print at all, but I'd still expect to see the brackets appear.
```python
from qiskit import QuantumCircuit, ClassicalRegister, QuantumRegister

cr = ClassicalRegister(2)
qc = QuantumCircuit(QuantumRegister(2), cr)
with qc.if_test((cr, 0)):
   pass
qc.draw()
```
```text
q3_0: ─

q3_1: ─

c6: 2/═
```
We then realised that this is because an empty control-flow block acts on zero qubits, so the problem is part of a larger issue around visualisation of zero-qubit instructions (#9962) and there's no action needed in this PR.

File: test/python/visualization/test_circuit_text_drawer.py (hunk 5 above)
Comment (enavarro51) [reply test/python/visualization/test_circuit_text_drawer.py:346]: You're right. I think I had an early problem with wire_order and cregbundle off, but that went away. I'll change it back.

Comment (enavarro51) [reply test/python/visualization/test_circuit_text_drawer.py:346]: Actually, I think there is a fundamental problem with wire_order changes to clbits and cregbundle on. How do you display the order if you have 3 qubits, and two 2-clbit registers with a wire order of [0, 2, 1, 6, 3, 4, 5]? I noticed that the tests for wire_order that are done in test_circuit_visualization only reorder the qubits. The reason this is an issue now is that with the control flow changes, the wire_map is used much more to access bit position than for example, self.qubits or self.clbits.

My suggestion would be to force cregbundle off if clbits are reordered.

Comment (enavarro51) [reply test/python/visualization/test_circuit_text_drawer.py:346]: Actually (actually) I just realized this is that crazy problem with the fact that the text drawer tests unlike the other drawers bypass the call to circuit_drawer which is where all the checks to force cregbundle false are done.

The reality is that this is a bad test, since cregbundle would normally be forced off by circuit_drawer before the call to the text drawer. So I say, either ditch the test or do what I did here, which is explicitly make cregbundle false.

File: qiskit/visualization/circuit/text.py (hunk 1 above)
Comment (enavarro51) [reply qiskit/visualization/circuit/text.py:47]: Noted.

File: qiskit/visualization/circuit/text.py (hunk 2 above)
Comment (enavarro51) [reply qiskit/visualization/circuit/text.py]: Done in https://github.com/Qiskit/qiskit/pull/10414/commits/9b82cd7387a174ec996ac853412e972faee3dcb9.

File: qiskit/visualization/circuit/text.py (hunk 3 above)
Comment (enavarro51) [reply qiskit/visualization/circuit/text.py:337]: After several tries, decided this is fraught with lots of formatting issues, so left it with left edges.

File: qiskit/visualization/circuit/text.py (hunk 4 above)
Comment (enavarro51) [reply qiskit/visualization/circuit/text.py]: Done in https://github.com/Qiskit/qiskit/pull/10414/commits/9b82cd7387a174ec996ac853412e972faee3dcb9.

File: test/python/visualization/test_circuit_text_drawer.py (hunk 5 above)
Comment (enavarro51) [reply test/python/visualization/test_circuit_text_drawer.py:346]: So left this with cregbundle false which is the only way it should be tested. To avoid problems in the future, I think it's worthwhile to switch to using circuit_drawer for text_drawer testing.

File: qiskit/visualization/circuit/text.py (hunk 3 above)
Comment (jakelishman) [reply qiskit/visualization/circuit/text.py:337]: Fine by me - it's all a bit of bike-shedding anyway.

File: test/python/visualization/test_circuit_text_drawer.py (hunk 5 above)
Comment (jakelishman) [reply test/python/visualization/test_circuit_text_drawer.py:346]: Yep, 100% agree.  Let's make that change in a separate PR with no feature changes, though, since it's a large-scale change to the testing structure.

File: qiskit/visualization/circuit/text.py [hunk 6]
```diff
@@ -1057,14 +1054,14 @@ def normalize_width(layer):
...
-    def controlled_wires(node, layer):
+    def controlled_wires(node, wire_map):
         """
         Analyzes the node in the layer and checks if the controlled arguments are in
         the box or out of the box.
 
         Args:
             node (DAGNode): node to analyse
-            layer (Layer): The layer in which the node is inserted.
+            wire_map (dict): map of qubits/clbits to position
```
Comment (jakelishman) [inline qiskit/visualization/circuit/text.py:1064]: Oh, you managed to get this working?  I'm pleased to see this - it looks generally more robust with the new idea of scopes.

**Review by jakelishman — APPROVED** [review]
This looks great - I'm super keen to get this merged and in so we can start using it, if you're also happy with it.
